app/code/core/Mage/Catalog/Model/Product/Action.php

```diff
@@ -46,7 +46,7 @@ protected function _construct()
     /**
      * Retrieve resource instance wrapper
      *
-     * @return Mage_Catalog_Model_Resource_Eav_Mysql4_Product_Action
+     * @inheritDoc
```

Comment:
This is a step backward for type inference. `@inheritDoc` will mean that all methods added by `Mage_Catalog_Model_Resource_Eav_Mysql4_Product_Action` will no longer be detected by the IDE because the parent method's docs say the return is `Mage_Core_Model_Mysql4_Abstract`.

Reply:
right. So what do you suggest? 
do 
```suggest
     * @return Mage_Catalog_Model_Resource_Product_Action
```
?
Probably other places need to be rechecked

Reply:
I'd leave it as-is ...

- `Mage_Catalog_Model_Resource_Eav_Mysql4_Product_Action` is an empty deprecated method
- `return parent::_getResource();` uses return type from parent DOCs ... so it seems correct
- `@return Mage_Core_Model_Mysql4_Abstract` should be changed to resource (later)

Problem is `Mage_Core_Model_Abstract::_getResource()` ... 
```
return Mage::getResourceSingleton($this->_resourceName);
```
... neither IDE nor phpstan can detect correct child-model of `Mage_Core_Model_Mysql4_Abstract`.